##### Copyright 2019 The TensorFlow Authors.

In [1]:
#@title Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

In [2]:
#@title MIT License
#
# Copyright (c) 2017 François Chollet
#
# Permission is hereby granted, free of charge, to any person obtaining a
# copy of this software and associated documentation files (the "Software"),
# to deal in the Software without restriction, including without limitation
# the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom the
# Software is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
# DEALINGS IN THE SOFTWARE.

# 使用 Keras 和 Tensorflow Hub 对电影评论进行文本分类

<table class="tfo-notebook-buttons" align="left">
  <td>     <a target="_blank" href="https://tensorflow.google.cn/tutorials/keras/text_classification_with_hub"><img src="https://tensorflow.google.cn/images/tf_logo_32px.png">在 TensorFlow.org 上查看</a>   </td>
  <td>     <a target="_blank" href="https://colab.research.google.com/github/tensorflow/docs-l10n/blob/master/site/zh-cn/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://tensorflow.google.cn/images/colab_logo_32px.png">在 Google Colab 中运行</a>   </td>
  <td>     <a target="_blank" href="https://github.com/tensorflow/docs-l10n/blob/master/site/zh-cn/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://tensorflow.google.cn/images/GitHub-Mark-32px.png">在 GitHub 上查看源代码</a>   </td>
  <td>     <a href="https://storage.googleapis.com/tensorflow_docs/docs-l10n/site/zh-cn/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://tensorflow.google.cn/images/download_logo_32px.png">下载笔记本</a>   </td>
  <td><a href="https://tfhub.dev/s?module-type=text-embedding"><img src="https://tensorflow.google.cn/images/hub_logo_32px.png">查看 TF Hub 模型</a></td>
</table>

此笔记本（notebook）使用评论文本将影评分为*积极（positive）*或*消极（nagetive）*两类。这是一个*二元（binary）*或者二分类问题，一种重要且应用广泛的机器学习问题。

本教程演示了使用 Tensorflow Hub 和 Keras 进行迁移学习的基本应用。

我们将使用包含 [Internet Movie Database](https://www.imdb.com/) 中的 50,000 条电影评论文本的 [IMDB 数据集](https://tensorflow.google.cn/api_docs/python/tf/keras/datasets/imdb)。先将这些评论分为两组，其中 25,000 条用于训练，另外 25,000 条用于测试。训练组和测试组是*均衡的*，也就是说其中包含相等数量的正面评价和负面评价。

此笔记本使用 [`tf.keras`](https://tensorflow.google.cn/guide/keras)（一个在 TensorFlow 中用于构建和训练模型的高级 API）和 [`tensorflow_hub`](https://tensorflow.google.cn/hub)（一个用于在一行代码中从 [TFHub](https://tfhub.dev) 加载训练模型的库）。有关使用 `tf.keras` 的更高级文本分类教程，请参阅 [MLCC 文本分类指南](https://developers.google.com/machine-learning/guides/text-classification/)。

In [1]:
!pip install tensorflow-hub
!pip install tensorflow-datasets

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 108.8/108.8 KB 350.0 kB/s eta 0:00:00a 0:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.2/4.2 MB 5.6 MB/s eta 0:00:0000:0100:010m
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 86.9/86.9 KB 3.4 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 48.8/48.8 KB 1.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 78.4/78.4 KB 2.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 241.5/241.5 KB 5.8 MB/s eta 0:00:0000:01
  Created wheel for promise: filename=promise-2.3-py3-none-any.whl size=21503 sha256=88016ac3ab0c04f6c363d7a7113e6fb634b2ef9ef23c1e6667bc45a51da3e245
  Stored in directory: /Users/junsircoding/Library/Caches/pip/wheels/54/aa/01/724885182f93150035a2a91bce34a12877e8067a97baaf5dc8
Successfully built promise


In [1]:
import numpy as np

import tensorflow as tf

!pip install tensorflow-hub
!pip install tfds-nightly
import tensorflow_hub as hub
import tensorflow_datasets as tfds

print("Version: ", tf.__version__)
print("Eager mode: ", tf.executing_eagerly())
print("Hub version: ", hub.__version__)
print("GPU is", "available" if tf.config.experimental.list_physical_devices("GPU") else "NOT AVAILABLE")

/Users/junsircoding/miniforge3/envs/tensor_env/lib/python3.8/site-packages/tqdm/auto.py:22: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


Version:  2.8.0
Eager mode:  True
Hub version:  0.12.0
GPU is available


## 下载 IMDB 数据集

IMDB数据集可以在 [Tensorflow 数据集](https://github.com/tensorflow/datasets)处获取。以下代码将 IMDB 数据集下载至您的机器（或 colab 运行时环境）中：

In [2]:
# Split the training set into 60% and 40% to end up with 15,000 examples
# for training, 10,000 examples for validation and 25,000 examples for testing.
train_data, validation_data, test_data = tfds.load(
    name="imdb_reviews", 
    split=('train[:60%]', 'train[60%:]', 'test'),
    as_supervised=True)

2022-05-01 18:23:52.133852: W tensorflow/core/platform/cloud/google_auth_provider.cc:184] All attempts to get a Google authentication bearer token failed, returning an empty token. Retrieving token from files failed with "NOT_FOUND: Could not locate the credentials file.". Retrieving token from GCE failed with "FAILED_PRECONDITION: Error executing an HTTP request: libcurl code 6 meaning 'Couldn't resolve host name', error details: Could not resolve host: metadata".
2022-05-01 18:24:53.137738: E tensorflow/core/platform/cloud/curl_http_request.cc:614] The transmission  of request 0x1327c3960 (URI: https://www.googleapis.com/storage/v1/b/tfds-data/o/dataset_info%2Fimdb_reviews%2Fplain_text%2F1.0.0?fields=size%2Cgeneration%2Cupdated) has been stuck at 0 of 0 bytes for 61 seconds and will be aborted. CURL timing information: lookup time: 0.001247 (No error), connect time: 0 (No error), pre-transfer time: 0 (No error), start-transfer time: 0 (No error)
2022-05-01 18:25:55.222824: E tensorfl

2022-05-01 18:37:32.595450: E tensorflow/core/platform/cloud/curl_http_request.cc:614] The transmission  of request 0x133969270 (URI: https://www.googleapis.com/storage/v1/b/tfds-data/o/datasets%2Fimdb_reviews%2Fplain_text%2F1.0.0?fields=size%2Cgeneration%2Cupdated) has been stuck at 0 of 0 bytes for 61 seconds and will be aborted. CURL timing information: lookup time: 0.001547 (No error), connect time: 0 (No error), pre-transfer time: 0 (No error), start-transfer time: 0 (No error)
2022-05-01 18:38:34.042561: E tensorflow/core/platform/cloud/curl_http_request.cc:614] The transmission  of request 0x10300c760 (URI: https://www.googleapis.com/storage/v1/b/tfds-data/o/datasets%2Fimdb_reviews%2Fplain_text%2F1.0.0?fields=size%2Cgeneration%2Cupdated) has been stuck at 0 of 0 bytes for 61 seconds and will be aborted. CURL timing information: lookup time: 0.002995 (No error), connect time: 0 (No error), pre-transfer time: 0 (No error), start-transfer time: 0 (No error)
2022-05-01 18:39:36.2520

Dataset imdb_reviews downloaded and prepared to ~/tensorflow_datasets/imdb_reviews/plain_text/1.0.0. Subsequent calls will reuse this data.
Metal device set to: Apple M1

systemMemory: 8.00 GB
maxCacheSize: 2.67 GB



## 探索数据

我们花一点时间来了解数据的格式。每个样本都是一个代表电影评论的句子和一个相应的标签。句子未经过任何预处理。标签是一个整数值（0 或 1），其中 0 表示负面评价，而 1 表示正面评价。

我们来打印下前十个样本。

In [6]:
train_examples_batch, train_labels_batch = next(iter(train_data.batch(10)))
train_examples_batch

2021-08-13 23:46:51.375995: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:185] None of the MLIR Optimization Passes are enabled (registered 2)


<tf.Tensor: shape=(10,), dtype=string, numpy=
array([b"This was an absolutely terrible movie. Don't be lured in by Christopher Walken or Michael Ironside. Both are great actors, but this must simply be their worst role in history. Even their great acting could not redeem this movie's ridiculous storyline. This movie is an early nineties US propaganda piece. The most pathetic scenes were those when the Columbian rebels were making their cases for revolutions. Maria Conchita Alonso appeared phony, and her pseudo-love affair with Walken was nothing but a pathetic emotional plug in a movie that was devoid of any real meaning. I am disappointed that there are movies like this, ruining actor's like Christopher Walken's good name. I could barely sit through it.",
       b'I have been known to fall asleep during films, but this is usually due to a combination of things including, really tired, being warm and comfortable on the sette and having just eaten a lot. However on this occasion I fell 

我们再打印下前十个标签。

In [7]:
train_labels_batch

<tf.Tensor: shape=(10,), dtype=int64, numpy=array([0, 0, 0, 1, 1, 1, 0, 0, 0, 0])>

## 构建模型

神经网络由堆叠的层来构建，这需要从三个主要方面来进行体系结构决策：

- 如何表示文本？
- 模型里有多少层？
- 每个层里有多少*隐层单元（hidden units）*？

本示例中，输入数据由句子组成。预测的标签为 0 或 1。

表示文本的一种方式是将句子转换为嵌入向量。使用一个预训练文本嵌入向量作为首层，这将具有三个优点：

- 不必担心文本预处理
- 可以从迁移学习中受益
- 嵌入具有固定长度，更易于处理

在本示例中，您使用来自 [TensorFlow Hub](https://tfhub.dev) 的 **预训练文本嵌入向量模型**，名称为 [google/nnlm-en-dim50/2](https://tfhub.dev/google/nnlm-en-dim50/2)。

本教程中还可以使用来自 TFHub 的许多其他预训练文本嵌入向量：

- [google/nnlm-en-dim128/2](https://tfhub.dev/google/nnlm-en-dim128/2) - 基于与 [google/nnlm-en-dim50/2](https://tfhub.dev/google/nnlm-en-dim50/2) 相同的数据并使用相同的 NNLM 架构进行训练，但具有更大的嵌入向量维度。更大维度的嵌入向量可以改进您的任务，但可能需要更长的时间来训练您的模型。
- [google/nnlm-en-dim128-with-normalization/2](https://tfhub.dev/google/nnlm-en-dim128-with-normalization/2) - 与 [google/nnlm-en-dim128/2](https://tfhub.dev/google/nnlm-en-dim128/2) 相同，但具有额外的文本归一化，例如移除标点符号。如果您的任务中的文本包含附加字符或标点符号，这会有所帮助。
- [google/universal-sentence-encoder/4](https://tfhub.dev/google/universal-sentence-encoder/4) - 一个可产生 512 维嵌入向量的更大模型，使用深度平均网络 (DAN) 编码器训练。

还有很多！在 TFHub 上查找更多[文本嵌入向量模型](https://tfhub.dev/s?module-type=text-embedding)。

让我们首先创建一个使用 Tensorflow Hub 模型嵌入（embed）语句的Keras层，并在几个输入样本中进行尝试。请注意无论输入文本的长度如何，嵌入（embeddings）输出的形状都是：`(num_examples, embedding_dimension)`。

In [8]:
embedding = "https://tfhub.dev/google/nnlm-en-dim50/2"
hub_layer = hub.KerasLayer(embedding, input_shape=[], 
                           dtype=tf.string, trainable=True)
hub_layer(train_examples_batch[:3])

<tf.Tensor: shape=(3, 50), dtype=float32, numpy=
array([[ 0.5423194 , -0.01190171,  0.06337537,  0.0686297 , -0.16776839,
        -0.10581177,  0.168653  , -0.04998823, -0.31148052,  0.07910344,
         0.15442258,  0.01488661,  0.03930155,  0.19772716, -0.12215477,
        -0.04120982, -0.27041087, -0.21922147,  0.26517656, -0.80739075,
         0.25833526, -0.31004202,  0.2868321 ,  0.19433866, -0.29036498,
         0.0386285 , -0.78444123, -0.04793238,  0.41102988, -0.36388886,
        -0.58034706,  0.30269453,  0.36308962, -0.15227163, -0.4439151 ,
         0.19462997,  0.19528405,  0.05666233,  0.2890704 , -0.28468323,
        -0.00531206,  0.0571938 , -0.3201319 , -0.04418665, -0.08550781,
        -0.55847436, -0.2333639 , -0.20782956, -0.03543065, -0.17533456],
       [ 0.56338924, -0.12339553, -0.10862677,  0.7753425 , -0.07667087,
        -0.15752274,  0.01872334, -0.08169781, -0.3521876 ,  0.46373403,
        -0.08492758,  0.07166861, -0.00670818,  0.12686071, -0.19326551,
 

现在让我们构建完整模型：

In [9]:
model = tf.keras.Sequential()
model.add(hub_layer)
model.add(tf.keras.layers.Dense(16, activation='relu'))
model.add(tf.keras.layers.Dense(1))

model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
keras_layer (KerasLayer)     (None, 50)                48190600  
_________________________________________________________________
dense (Dense)                (None, 16)                816       
_________________________________________________________________
dense_1 (Dense)              (None, 1)                 17        
Total params: 48,191,433
Trainable params: 48,191,433
Non-trainable params: 0
_________________________________________________________________


层按顺序堆叠以构建分类器：

1. 第一层是 TensorFlow Hub 层。此层使用预训练的 SaveModel 将句子映射到其嵌入向量。您使用的预训练文本嵌入向量模型 ([google/nnlm-en-dim50/2](https://tfhub.dev/google/nnlm-en-dim50/2)) 可将句子拆分为词例，嵌入每个词例，然后组合嵌入向量。生成的维度是：`(num_examples, embedding_dimension)`。对于此 NNLM 模型，`embedding_dimension` 是 50。
2. 该定长输出向量通过一个有 16 个隐层单元的全连接层（`Dense`）进行管道传输。
3. 最后一层与单个输出结点紧密相连。使用 `Sigmoid` 激活函数，其函数值为介于 0 与 1 之间的浮点数，表示概率或置信水平。

让我们编译模型。

### 损失函数与优化器

一个模型需要一个损失函数和一个优化器来训练。由于这是一个二元分类问题，并且模型输出 logit（具有线性激活的单一单元层），因此，我们将使用 `binary_crossentropy` 损失函数。

这并非损失函数的唯一选择，例如，您还可以选择 `mean_squared_error`。但是，一般来说，`binary_crossentropy` 更适合处理概率问题，它可以测量概率分布之间的“距离”，或者在我们的用例中，是指真实分布与预测值之间的差距。

稍后，当您探索回归问题（例如，预测房屋价格）时，您将看到如何使用另一个称为均方误差的损失函数。

现在，配置模型来使用优化器和损失函数：

In [10]:
model.compile(optimizer='adam',
              loss=tf.keras.losses.BinaryCrossentropy(from_logits=True),
              metrics=['accuracy'])

## 训练模型

使用包含 512 个样本的 mini-batch 对模型进行 10 个周期的训练，也就是在 `x_train` 和 `y_train` 张量中对所有样本进行 10 次迭代。在训练时，监测模型在验证集的 10,000 个样本上的损失和准确率：

In [11]:
history = model.fit(train_data.shuffle(10000).batch(512),
                    epochs=10,
                    validation_data=validation_data.batch(512),
                    verbose=1)

Epoch 1/10



 1/30 [>.............................] - ETA: 1:22 - loss: 0.7423 - accuracy: 0.4824


 3/30 [==>...........................] - ETA: 0s - loss: 0.7357 - accuracy: 0.4733  


 5/30 [====>.........................] - ETA: 0s - loss: 0.7270 - accuracy: 0.4844


 7/30 [======>.......................] - ETA: 0s - loss: 0.7177 - accuracy: 0.4841


10/30 [=========>....................] - ETA: 0s - loss: 0.7092 - accuracy: 0.4980


13/30 [============>.................] - ETA: 0s - loss: 0.7053 - accuracy: 0.4961


16/30 [===============>..............] - ETA: 0s - loss: 0.7003 - accuracy: 0.4991


19/30 [==================>...........] - ETA: 0s - loss: 0.6957 - accuracy: 0.5005


22/30 [=====================>........] - ETA: 0s - loss: 0.6908 - accuracy: 0.5059


25/30 [========================>.....] - ETA: 0s - loss: 0.6867 - accuracy: 0.5095


28/30 [===========================>..] - ETA: 0s - loss: 0.6823 - accuracy: 0.5168


30/30 [==============================] - 6s 119ms/step - loss: 0.6806 - accuracy: 0.5195 - val_loss: 0.6309 - val_accuracy: 0.6150


Epoch 2/10



 1/30 [>.............................] - ETA: 1:12 - loss: 0.6164 - accuracy: 0.6191


 3/30 [==>...........................] - ETA: 0s - loss: 0.6191 - accuracy: 0.6139  


 5/30 [====>.........................] - ETA: 0s - loss: 0.6150 - accuracy: 0.6309


 7/30 [======>.......................] - ETA: 0s - loss: 0.6083 - accuracy: 0.6392


 9/30 [========>.....................] - ETA: 0s - loss: 0.6035 - accuracy: 0.6363


12/30 [===========>..................] - ETA: 0s - loss: 0.5963 - accuracy: 0.6471


15/30 [==============>...............] - ETA: 0s - loss: 0.5918 - accuracy: 0.6510


18/30 [=================>............] - ETA: 0s - loss: 0.5886 - accuracy: 0.6528


21/30 [====================>.........] - ETA: 0s - loss: 0.5834 - accuracy: 0.6588


24/30 [=======================>......] - ETA: 0s - loss: 0.5771 - accuracy: 0.6683


27/30 [==========================>...] - ETA: 0s - loss: 0.5711 - accuracy: 0.6746


30/30 [==============================] - ETA: 0s - loss: 0.5663 - accuracy: 0.6782


30/30 [==============================] - 5s 97ms/step - loss: 0.5663 - accuracy: 0.6782 - val_loss: 0.5228 - val_accuracy: 0.7067


Epoch 3/10



 1/30 [>.............................] - ETA: 1:29 - loss: 0.4905 - accuracy: 0.7520


 3/30 [==>...........................] - ETA: 0s - loss: 0.4923 - accuracy: 0.7415  


 5/30 [====>.........................] - ETA: 0s - loss: 0.4804 - accuracy: 0.7523


 8/30 [=======>......................] - ETA: 0s - loss: 0.4739 - accuracy: 0.7690


11/30 [==========>...................] - ETA: 0s - loss: 0.4663 - accuracy: 0.7766


14/30 [=============>................] - ETA: 0s - loss: 0.4582 - accuracy: 0.7864


17/30 [================>.............] - ETA: 0s - loss: 0.4530 - accuracy: 0.7906


20/30 [===================>..........] - ETA: 0s - loss: 0.4485 - accuracy: 0.7943


23/30 [======================>.......] - ETA: 0s - loss: 0.4446 - accuracy: 0.7995


26/30 [=========================>....] - ETA: 0s - loss: 0.4396 - accuracy: 0.8007


29/30 [============================>.] - ETA: 0s - loss: 0.4343 - accuracy: 0.8027


30/30 [==============================] - 5s 52ms/step - loss: 0.4343 - accuracy: 0.8028 - val_loss: 0.4217 - val_accuracy: 0.8203


Epoch 4/10



 1/30 [>.............................] - ETA: 21s - loss: 0.3225 - accuracy: 0.8926


 3/30 [==>...........................] - ETA: 0s - loss: 0.3502 - accuracy: 0.8783 


 5/30 [====>.........................] - ETA: 0s - loss: 0.3389 - accuracy: 0.8844


 8/30 [=======>......................] - ETA: 0s - loss: 0.3416 - accuracy: 0.8770


11/30 [==========>...................] - ETA: 0s - loss: 0.3370 - accuracy: 0.8745


14/30 [=============>................] - ETA: 0s - loss: 0.3365 - accuracy: 0.8691


17/30 [================>.............] - ETA: 0s - loss: 0.3318 - accuracy: 0.8747


20/30 [===================>..........] - ETA: 0s - loss: 0.3287 - accuracy: 0.8778


23/30 [======================>.......] - ETA: 0s - loss: 0.3238 - accuracy: 0.8783


26/30 [=========================>....] - ETA: 0s - loss: 0.3197 - accuracy: 0.8793


29/30 [============================>.] - ETA: 0s - loss: 0.3149 - accuracy: 0.8818


30/30 [==============================] - 2s 55ms/step - loss: 0.3140 - accuracy: 0.8825 - val_loss: 0.3558 - val_accuracy: 0.8494


Epoch 5/10



 1/30 [>.............................] - ETA: 18s - loss: 0.2623 - accuracy: 0.9180


 3/30 [==>...........................] - ETA: 0s - loss: 0.2537 - accuracy: 0.9128 


 5/30 [====>.........................] - ETA: 0s - loss: 0.2525 - accuracy: 0.9152


 8/30 [=======>......................] - ETA: 0s - loss: 0.2507 - accuracy: 0.9167


11/30 [==========>...................] - ETA: 0s - loss: 0.2453 - accuracy: 0.9153


14/30 [=============>................] - ETA: 0s - loss: 0.2393 - accuracy: 0.9173


17/30 [================>.............] - ETA: 0s - loss: 0.2361 - accuracy: 0.9191


20/30 [===================>..........] - ETA: 0s - loss: 0.2320 - accuracy: 0.9207


23/30 [======================>.......] - ETA: 0s - loss: 0.2279 - accuracy: 0.9221


26/30 [=========================>....] - ETA: 0s - loss: 0.2281 - accuracy: 0.9214


29/30 [============================>.] - ETA: 0s - loss: 0.2263 - accuracy: 0.9225


30/30 [==============================] - 2s 53ms/step - loss: 0.2259 - accuracy: 0.9225 - val_loss: 0.3223 - val_accuracy: 0.8595


Epoch 6/10



 1/30 [>.............................] - ETA: 20s - loss: 0.1828 - accuracy: 0.9453


 3/30 [==>...........................] - ETA: 1s - loss: 0.1879 - accuracy: 0.9421 


 5/30 [====>.........................] - ETA: 0s - loss: 0.1820 - accuracy: 0.9398


 7/30 [======>.......................] - ETA: 0s - loss: 0.1779 - accuracy: 0.9436


10/30 [=========>....................] - ETA: 0s - loss: 0.1767 - accuracy: 0.9479


13/30 [============>.................] - ETA: 0s - loss: 0.1765 - accuracy: 0.9453


16/30 [===============>..............] - ETA: 0s - loss: 0.1737 - accuracy: 0.9452


19/30 [==================>...........] - ETA: 0s - loss: 0.1719 - accuracy: 0.9459


22/30 [=====================>........] - ETA: 0s - loss: 0.1684 - accuracy: 0.9479


25/30 [========================>.....] - ETA: 0s - loss: 0.1658 - accuracy: 0.9488


28/30 [===========================>..] - ETA: 0s - loss: 0.1653 - accuracy: 0.9482


30/30 [==============================] - 2s 54ms/step - loss: 0.1651 - accuracy: 0.9479 - val_loss: 0.3099 - val_accuracy: 0.8649


Epoch 7/10



 1/30 [>.............................] - ETA: 20s - loss: 0.1338 - accuracy: 0.9590


 3/30 [==>...........................] - ETA: 0s - loss: 0.1363 - accuracy: 0.9570 


 5/30 [====>.........................] - ETA: 0s - loss: 0.1346 - accuracy: 0.9574


 7/30 [======>.......................] - ETA: 0s - loss: 0.1291 - accuracy: 0.9623


 9/30 [========>.....................] - ETA: 0s - loss: 0.1302 - accuracy: 0.9618


12/30 [===========>..................] - ETA: 0s - loss: 0.1273 - accuracy: 0.9627


15/30 [==============>...............] - ETA: 0s - loss: 0.1260 - accuracy: 0.9618


18/30 [=================>............] - ETA: 0s - loss: 0.1243 - accuracy: 0.9638


21/30 [====================>.........] - ETA: 0s - loss: 0.1241 - accuracy: 0.9640


24/30 [=======================>......] - ETA: 0s - loss: 0.1222 - accuracy: 0.9651


27/30 [==========================>...] - ETA: 0s - loss: 0.1206 - accuracy: 0.9656


30/30 [==============================] - ETA: 0s - loss: 0.1202 - accuracy: 0.9654


30/30 [==============================] - 2s 53ms/step - loss: 0.1202 - accuracy: 0.9654 - val_loss: 0.3119 - val_accuracy: 0.8721


Epoch 8/10



 1/30 [>.............................] - ETA: 21s - loss: 0.1159 - accuracy: 0.9707


 3/30 [==>...........................] - ETA: 0s - loss: 0.1019 - accuracy: 0.9753 


 5/30 [====>.........................] - ETA: 0s - loss: 0.1012 - accuracy: 0.9730


 7/30 [======>.......................] - ETA: 0s - loss: 0.0967 - accuracy: 0.9763


 9/30 [========>.....................] - ETA: 0s - loss: 0.0944 - accuracy: 0.9757


12/30 [===========>..................] - ETA: 0s - loss: 0.0929 - accuracy: 0.9761


15/30 [==============>...............] - ETA: 0s - loss: 0.0898 - accuracy: 0.9775


18/30 [=================>............] - ETA: 0s - loss: 0.0881 - accuracy: 0.9785


21/30 [====================>.........] - ETA: 0s - loss: 0.0889 - accuracy: 0.9775


24/30 [=======================>......] - ETA: 0s - loss: 0.0883 - accuracy: 0.9779


27/30 [==========================>...] - ETA: 0s - loss: 0.0872 - accuracy: 0.9782


30/30 [==============================] - ETA: 0s - loss: 0.0864 - accuracy: 0.9785


30/30 [==============================] - 2s 51ms/step - loss: 0.0864 - accuracy: 0.9785 - val_loss: 0.3132 - val_accuracy: 0.8706


Epoch 9/10



 1/30 [>.............................] - ETA: 18s - loss: 0.0668 - accuracy: 0.9824


 3/30 [==>...........................] - ETA: 0s - loss: 0.0667 - accuracy: 0.9883 


 5/30 [====>.........................] - ETA: 0s - loss: 0.0698 - accuracy: 0.9875


 7/30 [======>.......................] - ETA: 0s - loss: 0.0673 - accuracy: 0.9891


10/30 [=========>....................] - ETA: 0s - loss: 0.0681 - accuracy: 0.9877


13/30 [============>.................] - ETA: 0s - loss: 0.0663 - accuracy: 0.9872


16/30 [===============>..............] - ETA: 0s - loss: 0.0644 - accuracy: 0.9874


19/30 [==================>...........] - ETA: 0s - loss: 0.0631 - accuracy: 0.9882


22/30 [=====================>........] - ETA: 0s - loss: 0.0624 - accuracy: 0.9885


25/30 [========================>.....] - ETA: 0s - loss: 0.0614 - accuracy: 0.9884


28/30 [===========================>..] - ETA: 0s - loss: 0.0611 - accuracy: 0.9881


30/30 [==============================] - 2s 55ms/step - loss: 0.0615 - accuracy: 0.9879 - val_loss: 0.3235 - val_accuracy: 0.8700


Epoch 10/10



 1/30 [>.............................] - ETA: 53s - loss: 0.0612 - accuracy: 0.9844


 3/30 [==>...........................] - ETA: 0s - loss: 0.0488 - accuracy: 0.9896 


 5/30 [====>.........................] - ETA: 0s - loss: 0.0497 - accuracy: 0.9906


 7/30 [======>.......................] - ETA: 0s - loss: 0.0467 - accuracy: 0.9919


10/30 [=========>....................] - ETA: 0s - loss: 0.0456 - accuracy: 0.9930


13/30 [============>.................] - ETA: 0s - loss: 0.0454 - accuracy: 0.9926


16/30 [===============>..............] - ETA: 0s - loss: 0.0448 - accuracy: 0.9924


19/30 [==================>...........] - ETA: 0s - loss: 0.0442 - accuracy: 0.9930


22/30 [=====================>........] - ETA: 0s - loss: 0.0439 - accuracy: 0.9934


25/30 [========================>.....] - ETA: 0s - loss: 0.0437 - accuracy: 0.9933


28/30 [===========================>..] - ETA: 0s - loss: 0.0439 - accuracy: 0.9929


30/30 [==============================] - 3s 53ms/step - loss: 0.0439 - accuracy: 0.9929 - val_loss: 0.3368 - val_accuracy: 0.8675


## 评估模型

我们来看下模型的表现如何。将返回两个值。损失值（loss）（一个表示误差的数字，值越低越好）与准确率（accuracy）。

In [12]:
results = model.evaluate(test_data.batch(512), verbose=2)

for name, value in zip(model.metrics_names, results):
  print("%s: %.3f" % (name, value))

49/49 - 2s - loss: 0.3661 - accuracy: 0.8526


loss: 0.366
accuracy: 0.853


这种十分朴素的方法得到了约 87% 的准确率（accuracy）。若采用更好的方法，模型的准确率应当接近 95%。

## 进一步阅读

- 有关处理字符串输入的更通用方式以及对训练过程中准确率和损失进度的更详细分析，请参阅[使用预处理文本的文本分类](./text_classification.ipynb)教程。
- 尝试更多使用来自 TFHub 的训练模型的[文本相关教程](https://tensorflow.google.cn/hub/tutorials#text-related-tutorials)。